nornir/plugins/inventory/ansible.py

```diff
@@ -242,24 +253,178 @@ def load_hosts_file(self) -> None:
+                ]
+            )
+        elif inv.is_file():
+            possible_sources.append(str(inv.resolve()))
+
+    valid_sources: List[AnsibleParser] = []
+    for possible_source in possible_sources:
         try:
-            parser = YAMLParser(hostsfile)
+            parser: AnsibleParser = INIParser(possible_source)
+            valid_sources.append(parser)
+            continue
+        except cp.Error:
+            logger.info(
+                "AnsibleInventory: file %r is not INI file, moving to next parser...",
+                possible_source,
+            )
+        try:
+            parser = YAMLParser(possible_source)
+            valid_sources.append(parser)
+            continue
         except (ScannerError, ComposerError):
-            logger.error("AnsibleInventory: file %r is not INI or YAML file", hostsfile)
-            raise
+            logger.info(
+                "AnsibleInventory: file %r is not YAML file, moving to next parser...",
+                possible_source,
+            )
+        try:
+            parser = ScriptParser(possible_source)
+            valid_sources.append(parser)
+            continue
+        except (TypeError, OSError, JSONDecodeError) as e:
+            logger.info(
+                "AnsibleInventory: file %r is not executable Python file. "
+                "Error: %r no more parsers to try...",
+                possible_source,
+                e,
+            )
 
-    parser.parse()
+    if not valid_sources:
+        raise NornirNoValidInventoryError(
+            f"AnsibleInventory: no valid inventory source(s) to parse. Tried: {possible_sources}"
+        )
+
+    hosts: Dict[str, Any] = {}
+    groups: Dict[str, Any] = {}
+    defaults: Dict[str, Any] = {}
+
+    for source in valid_sources:
+        source.parse()
+        hosts = combine_vars(hosts, source.hosts, hash_behavior)
+        groups = combine_vars(groups, source.groups, hash_behavior)
+        defaults = combine_vars(defaults, source.defaults, hash_behavior)
+
+    return hosts, groups, defaults
+
+
+def combine_vars(
+    a: Dict[str, Any], b: Dict[str, Any], hash_behavior: str
+) -> Dict[str, Any]:
+    if hash_behavior == "merge":
+        return merge_hash(a, b)
+    return replace_hash(a, b)
+
+
+def merge_hash(a: Dict[str, Any], b: Union[Dict[str, Any]]) -> Dict[str, Any]:
+    if a == {} or a == b:
+        return b.copy()
+    result = a.copy()
+    for k, v in b.items():
+        if (
+            k in result
+            and isinstance(result[k], MutableMapping)
+            and isinstance(v, MutableMapping)
+        ):
+            new_v: Dict[str, Any] = dict(v)
+            result[k] = merge_hash(result[k], new_v)
+        else:
+            result[k] = v
+    return result
 
-    return parser.hosts, parser.groups, parser.defaults
+
+def replace_hash(a: Dict[str, Any], b: Dict[str, Any]) -> Dict[str, Any]:
+    result = a.copy()
```

Comment:
shouldn't we be using deepcopy to avoid copying references at the second level?

Reply:
Yes? I kinda thought the same, but this was pretty much lifted straight from Ansible source [here](https://github.com/ansible/ansible/blob/b31b5d105a84aa14b17302d2778c7606c78912bf/lib/ansible/utils/vars.py#L95). In the testing I've done so far (using this for some tests at work with a fairly big inventory too) I've not seen any issue.... but that doesn't mean its right!

Replacing copy w/ deep copy on the merge_hash and replace_hash functions works as expected in tests so I can update that if ya want, just let me know!